src/lib.rs

```diff
@@ -14,39 +24,184 @@ fn process_instruction(
+                    funder_info.key,
+                    program_token_info.key,
+                    1.max(rent.minimum_balance(spl_token::state::Account::get_packed_len())),
+                    spl_token::state::Account::get_packed_len() as u64,
+                    &spl_token::id(),
+                ),
+                &[
+                    funder_info.clone(),
+                    program_token_info.clone(),
+                    system_program_info.clone(),
+                ],
+                &[&program_token_signer_seeds],
+            )?;
+
+            info!("Initializing program token account");
+            invoke(
+                &spl_token::instruction::initialize_account(
+                    &spl_token::id(),
+                    program_token_info.key,
+                    mint_info.key,
+                    program_token_info.key, // token owner is also `program_token` address
+                )?,
+                &[
+                    program_token_info.clone(),
+                    spl_token_program_info.clone(),
+                    rent_sysvar_info.clone(),
+                    mint_info.clone(),
+                ],
+            )?;
+            Ok(())
+        }
+        Some(1) => {
+            info!("Close program token account...");
+            let funder_info = next_account_info(account_info_iter)?;
+            let spl_token_program_info = next_account_info(account_info_iter)?;
+
+            invoke_signed(
+                &spl_token::instruction::close_account(
+                    &spl_token::id(),
+                    program_token_info.key,
+                    funder_info.key,
+                    program_token_info.key, // token owner is also `program_token` address
+                    &[],
+                )
+                .expect("close_account"),
+                &[
+                    funder_info.clone(),
+                    spl_token_program_info.clone(),
+                    program_token_info.clone(),
+                ],
+                &[&program_token_signer_seeds],
+            )
+        }
+        _ => {
+            info!("Error: Unsupported instruction");
+            Err(ProgramError::InvalidInstructionData)
+        }
+    }
 }
 
 #[cfg(test)]
 mod test {
+    #![cfg(feature = "test-bpf")]
+
     use super::*;
     use assert_matches::*;
-    use solana_program::instruction::{AccountMeta, Instruction};
+    use solana_program::{
+        instruction::{AccountMeta, Instruction},
+        sysvar,
+    };
     use solana_program_test::*;
     use solana_sdk::{signature::Signer, transaction::Transaction};
 
+    fn program_test(program_id: Pubkey) -> ProgramTest {
+        let mut pc = ProgramTest::new(
+            "bpf_program_template",
+            program_id,
+            processor!(process_instruction),
+        );
+
+        // Add SPL Token program
+        pc.add_program(
+            "spl_token",
+            spl_token::id(),
```

Comment:
i'm still confused by this `pc.add_program()` function, it seems like it's not even being used. i've browsed around through the repo and i can't really find precisely what `spl_token::id()` is, though i'm pretty sure it's just the spl_token's account address that we've added to `ProgramTest`

what is `spl_token::id()` here? and further, how is that the program itself and also the tests (see line 56) can simply use `spl_token::id()`?

going back to my confusion of `pc.add_program()`... on line 145, you've just called `let program_id = Pubkey::new_unique();` that returns a unique account pubkey, shouldn't the spl_token program we added to the program test be the program_id we're using in our tests?

Reply:
`spl_token::id()` is actually this: https://github.com/solana-labs/solana-program-library/blob/master/token/program/src/lib.rs#L29

> going back to my confusion of pc.add_program()... on line 145, you've just called let program_id = Pubkey::new_unique(); that returns a unique account pubkey, shouldn't the spl_token program we added to the program test be the program_id we're using in our tests?

`let program_id = Pubkey::new_unique()` is the program id of bpf_program_template, the actual program that was written here.  `spl_token::id()` is the program id of the SPL Token program, which is invoked by bpf_program_template